stablehlo/reference/NumPy.cpp

```diff
@@ -0,0 +1,323 @@
+/* Copyright 2023 The StableHLO Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "stablehlo/reference/NumPy.h"
+
+#include <fstream>
+#include <regex>
+
+#include "llvm/Support/Errc.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/raw_ostream.h"
+#include "mlir/Dialect/Quant/QuantTypes.h"
+#include "mlir/IR/BuiltinTypeInterfaces.h"
+#include "mlir/IR/Types.h"
+
+namespace mlir {
+namespace stablehlo {
+namespace numpy {
+namespace {
+
+constexpr char kMagicString[] = "\x93NUMPY";
```

Comment:
While I do like having less dependencies, this looks like it may end up having extra maintenance overhead given the versioning.

I'm curious of your thoughts - is there a library we should be taking a 3p dependency on for this? I tried a quick search and it doesn't seem like there's a standard C++ library.

Reply:
There is https://github.com/rogersce/cnpy, but no standard libraries to my knowledge. That being said, the numpy versioning has been quite stable, and isn't a fast moving target by any means (i.e. there's only been a couple of revisions since it was released.